config/parse.py

```diff
@@ -267,6 +267,18 @@ def __init__(self, config_file, verbose=False):
         self.caches = {k:v for k,v in self.caches.items() if k != 'DRAM'}
 
         self.pmem = config_file.get('physical_memory', {})
+        
+        #this allows frequency to be specified instead of data rate or vice-versa for DRAM
+        if('frequency' in self.pmem.keys()):
+            self.pmem['data_rate'] = self.pmem['frequency']
+            self.pmem['frequency'] = self.pmem['frequency']/2
+        elif('data_rate' in self.pmem.keys()):
+            self.pmem['frequency'] = self.pmem['data_rate']/2
+        
+        if('columns' in self.pmem.keys()):
+            print('W: physical_memory, "columns" is deprecated, use "bank_columns" instead (set to columns*8)')
+        if('rows' in self.pmem.keys()):
+            print('W: physical_memory, "rows" is deprecated, use "bank_rows" instead')
```

Comment:
I think you should use [this](https://github.com/ChampSim/ChampSim/blob/c90724c44405c1a553e2d66036c666fd107aea9a/config/parse.py#L159)

Reply:
Functions are overrated. Think about how much compute is wasted on jumps 🧠. In a performative scripting language like Python, you need to consider these things 🧐

Reply:
I wish there were some microarchitectural way we could predict the destination of jumps. It would probably enable testable, reusable code.

Reply:
I don't think this function plays nice with pmem. I have tried for the last 30 minutes to get it to work, and it won't

Reply:
Figured it out. I guess util.combine_named doesn't like scalar elements.